File: 221.maximal-square.py
#
# @lc app=leetcode id=221 lang=python3
#
# [221] Maximal Square
#
from typing import *
# @lc code=start
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        if not len(matrix):
            return 0
        row_count = len(matrix)
        column_count = len(matrix[0])
        square = [[0]*column_count for row in range(row_count)]
        for row_index in range(row_count):
            for column_index in range(column_count):
                if matrix[row_index][column_index] == '1':
                    if row_index > 0 and column_index > 0:          
                        square[row_index][column_index] = min(square[row_index-1][column_index-1],square[row_index][column_index-1],square[row_index-1][column_index]) +1
                    else:
                        square[row_index][column_index] = 1

        return max([max(row)for row in square]) ** 2
# @lc code=end

if __name__ == "__main__":
    matrix = [["1","1","1","1","0"],["1","1","1","1","0"],["1","1","1","1","1"],["1","1","1","1","1"],["0","0","1","1","1"]]
    # print(Solution().maximalSquare([["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]))
    print(Solution().maximalSquare(matrix))

File: golang/221.maximal-square.go
/*
 * @lc app=leetcode.cn id=221 lang=golang
 *
 * [221] Maximal Square
 */
package main

// @lc code=start
func maximalSquare(matrix [][]byte) int {
	if len(matrix) == 0 {
		return 0
	}

	dp := make([][]int, len(matrix))
	for i := 0; i < len(matrix); i++ {
		dp[i] = make([]int, len(matrix[0]))
	}
	maxSide := 0
	for i := 0; i < len(matrix); i++ {
		for j := 0; j < len(matrix[0]); j++ {
			if matrix[i][j] == '0' {
				dp[i][j] = 0
			} else {
				if i == 0 || j == 0 {
					dp[i][j] = 1
				} else {
					dp[i][j] = minArray([]int{dp[i-1][j-1], dp[i][j-1], dp[i-1][j]}) + 1
				}
			}
			if maxSide < dp[i][j] {
				maxSide = dp[i][j]
			}
		}
	}
	return maxSide * maxSide
}
func minArray(nums []int) int {
	ret := nums[0]
	for _, num := range nums {
		if ret > num {
			ret = num
		}
	}
	return ret
}
func maxArray(nums []int) int {
	ret := nums[0]
	for _, num := range nums {
		if ret < num {
			ret = num
		}
	}
	return ret
}

// @lc code=end
